t2548: fix(claim-task-id): append TODO entry after verified issue creation to prevent orphans

## What

Fix orphan-task-id bug: `claim-task-id.sh` creates the GitHub issue (via `_try_issue_sync_delegation` OR the bare `gh issue create` fallback) but never writes a TODO.md entry when the task ID is not already present in TODO.md. Silent gap leaves issue on GitHub with no TODO tracking.

## Why

Root cause verified 2026-04-20:

- `issue-sync-helper.sh:728-733` (`_push_process_task`) — when `target_task` is passed but the task ID is absent from TODO.md, it prints `[WARN] Task <tid> not found in TODO.md` and returns 0 without echoing `CREATED #N`.
- `claim-task-id.sh:1015-1033` (`_try_issue_sync_delegation`) parses stdout for `Created #NNN` — no match → returns 1.
- Bare fallback (`claim-task-id.sh:1270-1318`) creates the issue and returns `$issue_num`, but never appends to TODO.md. `add_gh_ref_to_todo` (`issue-sync-lib.sh:1280`) is a MODIFY-existing-line helper.
- `claim-task-id.sh` has zero TODO.md append logic anywhere in the allocation path.

Evidence: 4/4 orphans from a single ILDS session 2026-04-20 (GH#555-558, t135-t138). `johnwaldo/ilds` has 9 open orphans in the current corpus (plus 18 closed). Silently leaking across pulse-enabled repos for months.

## How

Option A (append-after-verified-creation):

- Add `_ensure_todo_entry_written` helper in `claim-task-id.sh`. Called from `create_github_issue` AFTER `$issue_num` is resolved on both paths (delegation success and bare-fallback).
- Idempotent: if TODO.md already has `- [.] <task_id>` (normal `/new-task` flow), no-op (delegation path already called `add_gh_ref_to_todo`).
- Otherwise: append `- [ ] <task_id> <description> <tags-from-labels> ref:GH#<num>` under the existing `## Tasks` section (or end-of-file fallback if no such heading).
- Strong invariant: "TODO entry exists ⟹ GitHub issue exists" — no orphan TODO-without-ref intermediate state possible.

Files to modify:

- EDIT: `.agents/scripts/claim-task-id.sh` — add `_ensure_todo_entry_written` helper + call sites in `create_github_issue` (two paths: delegation success and bare fallback)
- NEW: `.agents/scripts/tests/test-claim-task-id-no-orphan.sh` — regression harness covering repro scenario
- NEW: `.agents/scripts/backfill-orphan-todos.sh` — helper to annotate existing orphans in a repo (tests against `johnwaldo/ilds` corpus, 9 open orphans)

## Acceptance

- [ ] `claim-task-id.sh --title "..." --description "..."` creates issue AND appends TODO entry with `ref:GH#NNN` in a single logical operation (no orphans)
- [ ] `issue-sync-helper.sh pull` on a fresh repo shows zero new orphans after a batch of 3+ `claim-task-id.sh` runs
- [ ] No regression in CI bulk push path (`.github/workflows/issue-sync.yml`)
- [ ] Regression test `tests/test-claim-task-id-no-orphan.sh` passes
- [ ] Backfill helper or documented one-liner to annotate existing orphans
- [ ] Memory lesson stored: `claim-task-id + issue-sync-helper push contract mismatch → orphans`

## Session Origin

interactive — Marcus requested framework bug fix, verified root cause in session 2026-04-20. Option A selected over B/C for smallest blast radius + strongest invariant.

## Files to modify

- EDIT: `.agents/scripts/claim-task-id.sh:1210-1320` (`create_github_issue`) — invoke new helper on both success paths
- NEW: `.agents/scripts/tests/test-claim-task-id-no-orphan.sh`
- NEW: `.agents/scripts/backfill-orphan-todos.sh`

<!-- aidevops:sig -->
---
[aidevops.sh](https://aidevops.sh) v3.8.85 plugin for [OpenCode](https://opencode.ai) v1.14.19 with claude-opus-4-7 spent 2h 14m and 16,797 tokens on this with the user in an interactive session.

<!-- t2118:prior-worker-diff:PR20184 -->
## Prior worker attempt (PR #20184, closed CONFLICTING)

The following diff was produced by the prior worker before the PR was closed
due to merge conflicts. The next worker should review this diff and
rebase/apply it rather than re-deriving the solution from scratch.

<details><summary>Diff from PR #20184 (click to expand)</summary>

```diff
diff --git a/.agents/scripts/backfill-orphan-todos.sh b/.agents/scripts/backfill-orphan-todos.sh
new file mode 100644
index 0000000000..a6dbbbe242
--- /dev/null
+++ b/.agents/scripts/backfill-orphan-todos.sh
@@ -0,0 +1,244 @@
+#!/usr/bin/env bash
+# SPDX-License-Identifier: MIT
+# SPDX-FileCopyrightText: 2025-2026 Marcus Quinn
+#
+# backfill-orphan-todos.sh — one-shot repair tool for the t2548 orphan bug.
+#
+# Scans a repo's GitHub issues for "orphans" — issues whose title starts
+# with `tNNN: ` but whose `tNNN` ID has NO matching entry in TODO.md.
+# These are the pre-fix leakage from `claim-task-id.sh` creating issues
+# without writing TODO.md (fixed in t2548 / GH#20180).
+#
+# Behaviour:
+#   - `list` (default): print each orphan as `tNNN<TAB>GH#<num><TAB>title`
+#   - `annotate`: append a `- [ ] tNNN <title> ref:GH#<num>` entry to
+#     the `## Backlog` section of TODO.md for every open orphan.
+#     Idempotent — skips IDs that already exist in TODO.md.
+#     WRITES TODO.md in place; run inside a worktree, commit, PR.
+#
+# Scope:
+#   - Open issues only by default. Use `--include-closed` to also
+#     annotate orphans from closed issues (for audit-trail completeness).
+#   - Only issues whose title matches `^t[0-9]+: `. Issues using
+#     `GH#NNN:` or other prefix formats are not considered orphans.
+#
+# Usage:
+#   backfill-orphan-todos.sh list [--repo <path>] [--include-closed]
+#   backfill-orphan-todos.sh annotate [--repo <path>] [--include-closed] [--dry-run]
+#
+# Exit codes:
+#   0  - success (or no orphans found)
+#   1  - error (missing gh, no TODO.md, etc.)
+#
+# Cross-references:
+#   - t2548: root-cause fix (claim-task-id.sh::_ensure_todo_entry_written)
+#   - Evidence corpus: johnwaldo/ilds (9 open orphans 2026-04-20)
+
+set -uo pipefail
+
+SCRIPT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)" || exit 1
+
+# shellcheck source=shared-constants.sh
+if [[ -f "${SCRIPT_DIR}/shared-constants.sh" ]]; then
+	source "${SCRIPT_DIR}/shared-constants.sh"
+fi
+
+# Fallback log/print helpers if shared-constants.sh isn't available.
+if ! declare -F print_info >/dev/null 2>&1; then
+	print_info() { printf '[INFO] %s\n' "$*" >&2; }
+fi
+if ! declare -F print_warning >/dev/null 2>&1; then
+	print_warning() { printf '[WARN] %s\n' "$*" >&2; }
+fi
+if ! declare -F print_error >/dev/null 2>&1; then
+	print_error() { printf '[ERROR] %s\n' "$*" >&2; }
+fi
+
+usage() {
+	sed -n '7,32p' "$0" | sed 's/^# \{0,1\}//'
+	return 0
+}
+
+COMMAND=""
+REPO_PATH=""
+INCLUDE_CLOSED=false
+DRY_RUN=false
+
+parse_args() {
+	local cmd="$1"
+	COMMAND="${cmd:-list}"
+	shift || true
+	while [[ $# -gt 0 ]]; do
+		local flag="$1"
+		shift
+		case "$flag" in
+		--repo)
+			local val="$1"
+			REPO_PATH="$val"
+			shift
+			;;
+		--include-closed)
+			INCLUDE_CLOSED=true
+			;;
+		--dry-run)
+			DRY_RUN=true
+			;;
+		-h | --help)
+			usage
+			exit 0
+			;;
+		*)
+			print_error "unknown flag: $flag"
+			usage
+			exit 1
+			;;
+		esac
+	done
+	return 0
+}
+
+parse_args "$@"
+
+if [[ -z "$REPO_PATH" ]]; then
+	REPO_PATH="$(pwd)"
+fi
+
+if ! command -v gh >/dev/null 2>&1; then
+	print_error "gh CLI not found"
+	exit 1
+fi
+
+TODO_FILE="${REPO_PATH}/TODO.md"
+if [[ ! -f "$TODO_FILE" ]]; then
+	print_error "no TODO.md at ${REPO_PATH}"
+	exit 1
+fi
+
+SLUG=$(git -C "$REPO_PATH" remote get-url origin 2>/dev/null \
+	| sed 's|.*github\.com[:/]||;s|\.git$||' || true)
+if [[ -z "$SLUG" ]]; then
+	print_error "could not resolve GitHub slug from origin remote in ${REPO_PATH}"
+	exit 1
+fi
+
+# Fetch issue titles + numbers (tNNN-prefixed only).
+STATE_FILTER="open"
+if [[ "$INCLUDE_CLOSED" == "true" ]]; then
+	STATE_FILTER="all"
+fi
+
+print_info "scanning ${SLUG} (${STATE_FILTER}) for tNNN: orphans..."
+
+ISSUES_JSON=$(gh issue list --repo "$SLUG" --state "$STATE_FILTER" \
+	--search "t in:title" --limit 1000 \
+	--json number,title 2>/dev/null || echo "[]")
+
+# Parse with jq: keep only titles matching ^t[0-9]+:
+CANDIDATES=$(printf '%s' "$ISSUES_JSON" | jq -r '
+	.[]
+	| select(.title | test("^t[0-9]+:"))
+	| [
+		(.title | capture("^(?<id>t[0-9]+):") | .id),
+		(.number | tostring),
+		.title
+	]
+	| @tsv
+' 2>/dev/null || echo "")
+
+if [[ -z "$CANDIDATES" ]]; then
+	print_info "no tNNN-prefixed issues found"
+	exit 0
+fi
+
+# Load TODO.md IDs once (outside code fences, via literal scan is fine for
+# this audit — false-positives inside code fences just mean we SKIP a real
+# orphan, which is safe).
+declare -a ORPHANS=()
+
+while IFS=$'\t' read -r task_id issue_num title; do
+	[[ -z "$task_id" || -z "$issue_num" ]] && continue
+	# Check whether TODO.md has a `- [.] tNNN ` entry outside code fences
+	# (rough check: grep the whole file — if TODO.md has tNNN anywhere
+	# as a task line, assume it's tracked).
+	if grep -qE "^[[:space:]]*- \[.\] ${task_id}( |\$)" "$TODO_FILE"; then
+		continue
+	fi
+	ORPHANS+=("${task_id}"$'\t'"${issue_num}"$'\t'"${title}")
+done <<<"$CANDIDATES"
+
+ORPHAN_COUNT=${#ORPHANS[@]}
+
+if [[ "$ORPHAN_COUNT" -eq 0 ]]; then
+	print_info "no orphans found in ${SLUG}"
+	exit 0
+fi
+
+case "$COMMAND" in
+list)
+	print_info "${ORPHAN_COUNT} orphan(s) in ${SLUG}:"
+	printf '%s\n' "${ORPHANS[@]}"
+	exit 0
+	;;
+annotate)
+	print_info "${ORPHAN_COUNT} orphan(s) to annotate in ${TODO_FILE}"
+	if [[ "$DRY_RUN" == "true" ]]; then
+		print_info "[DRY-RUN] would append:"
+		for row in "${ORPHANS[@]}"; do
+			IFS=$'\t' read -r task_id issue_num title <<<"$row"
+			desc="${title#"${task_id}": }"
+			printf '  - [ ] %s %s ref:GH#%s\n' "$task_id" "$desc" "$issue_num"
+		done
+		exit 0
+	fi
+
+	# Find ## Backlog insertion point.
+	BACKLOG_LINE=$(grep -nE '^## Backlog[[:space:]]*$' "$TODO_FILE" 2>/dev/null | head -1 | cut -d: -f1 || true)
+
+	TMP=$(mktemp)
+	trap 'rm -f "$TMP"' EXIT
+
+	if [[ -n "$BACKLOG_LINE" ]]; then
+		NEXT_HEADING=$(awk -v s="$BACKLOG_LINE" 'NR > s && /^## / {print NR; exit}' "$TODO_FILE" 2>/dev/null || true)
+		# Build the block of lines to insert.
+		BLOCK=""
+		for row in "${ORPHANS[@]}"; do
+			IFS=$'\t' read -r task_id issue_num title <<<"$row"
+			desc="${title#"${task_id}": }"
+			BLOCK+="- [ ] ${task_id} ${desc} ref:GH#${issue_num}"$'\n'
+		done
+
+		if [[ -z "$NEXT_HEADING" ]]; then
+			# Backlog is last — append block at end of file.
+			cat "$TODO_FILE" >"$TMP"
+			# Ensure trailing newline, then block.
+			[[ -n "$(tail -c 1 "$TMP" 2>/dev/null)" ]] && printf '\n' >>"$TMP"
+			printf '%s' "$BLOCK" >>"$TMP"
+		else
+			# Insert block immediately before NEXT_HEADING.
+			awk -v nh="$NEXT_HEADING" -v block="$BLOCK" '
+				NR == nh { printf "%s\n", block }
+				{ print }
+			' "$TODO_FILE" >"$TMP"
+		fi
+	else
+		# No Backlog — append with heading.
+		cat "$TODO_FILE" >"$TMP"
+		printf '\n## Backlog (t2548 backfill)\n\n' >>"$TMP"
+		for row in "${ORPHANS[@]}"; do
+			IFS=$'\t' read -r task_id issue_num title <<<"$row"
+			desc="${title#"${task_id}": }"
+			printf -- '- [ ] %s %s ref:GH#%s\n' "$task_id" "$desc" "$issue_num" >>"$TMP"
+		done
+	fi
+
+	mv "$TMP" "$TODO_FILE"
+	trap - EXIT
+	print_info "annotated ${ORPHAN_COUNT} orphan(s) into ${TODO_FILE} — review with 'git diff' and commit"
+	exit 0
+	;;
+*)
+	print_error "unknown command: ${COMMAND}"
+	usage
+	exit 1
+	;;
+esac
diff --git a/.agents/scripts/claim-task-id.sh b/.agents/scripts/claim-task-id.sh
index 129391f643..81ce7427bc 100755
--- a/.agents/scripts/claim-task-id.sh
+++ b/.agents/scripts/claim-task-id.sh
@@ -1203,16 +1203,164 @@ _compose_issue_body() {
 	return 0
 }
 
+# t2548: Ensure TODO.md has an entry for the task after a verified issue
+# creation. Closes the orphan-task-id gap: `claim-task-id.sh` used to create
+# GitHub issues (via delegation OR bare `gh issue create` fallback) without
+# ever appending a TODO entry when the task ID wasn't already present —
+# `_push_process_task` in `issue-sync-helper.sh` silently returns when
+# `target_task` is not found in TODO.md, leaving the issue untracked.
+#
+# Fix strategy (Option A — append-after-verified-creation):
+#   - Idempotent: if `- [.] <task_id>` already in TODO.md, delegate to
+#     `add_gh_ref_to_todo` (which is also idempotent on the ref stamp).
+#   - Otherwise append `- [ ] <task_id> <description> <tags> ref:GH#<num>`
+#     to the end of the `## Backlog` section (falls back to EOF if absent).
+#   - Labels with `status:`, `tier:`, `origin:`, `dispatched:`, `implemented:`
+#     prefixes are skipped — they are not tags in the TODO-file format.
+#
+# Only fires when the issue creation verifiably succeeded, so the invariant
+# "TODO entry exists ⟹ GitHub issue exists" is preserved.
+#
+# Arguments:
+#   $1 - task_id (e.g. t2548)
+#   $2 - issue_number (integer, no `#`)
+#   $3 - description (single-line; newlines/tabs squashed to spaces)
+#   $4 - labels (CSV from --labels; empty allowed)
+#   $5 - repo_path
+#
+# Returns 0 always (non-fatal; the issue is already created).
+# _insert_todo_line FILE LINE
+# Write LINE into FILE at the end of the ## Backlog section, or at EOF.
+# Extracted from _ensure_todo_entry_written to keep function bodies <=100 lines.
+_insert_todo_line() {
+	local todo_file="$1"
+	local todo_line="$2"
+
+	local backlog_start
+	backlog_start=$(grep -nE '^## Backlog[[:space:]]*$' "$todo_file" 2>/dev/null | head -1 | cut -d: -f1 || true)
+
+	local tmp
+	tmp=$(mktemp)
+	if [[ -n "$backlog_start" ]]; then
+		local next_heading
+		next_heading=$(awk -v s="$backlog_start" 'NR > s && /^## / {print NR; exit}' "$todo_file" 2>/dev/null || true)
+		if [[ -z "$next_heading" ]]; then
+			# Backlog is the last section — insert before trailing blank lines.
+			awk -v line="$todo_line" '
+				{ lines[NR] = $0 }
+				END {
+					last = NR
+					while (last > 0 && lines[last] == "") last--
+					for (i = 1; i <= last; i++) print lines[i]
+					print line
+					for (i = last + 1; i <= NR; i++) print lines[i]
+				}
+			' "$todo_file" >"$tmp"
+		else
+			# Insert just before the next heading.
+			awk -v nh="$next_heading" -v line="$todo_line" '
+				NR == nh { print line; print "" }
+				{ print }
+			' "$todo_file" >"$tmp"
+		fi
+	else
+		# No Backlog section — append at EOF.
+		cat "$todo_file" >"$tmp"
+		printf '\n%s\n' "$todo_line" >>"$tmp"
+	fi
+
+	if [[ -s "$tmp" ]]; then
+		mv "$tmp" "$todo_file"
+	else
+		rm -f "$tmp"
+	fi
+	return 0
+}
+
+# _ensure_todo_entry_written TASK_ID ISSUE_NUM DESCRIPTION LABELS REPO_PATH
+# Idempotently appends a TODO.md entry after verified GitHub issue creation.
+# Fixes the t2548 orphan gap where both create_github_issue() paths created
+# issues without writing a corresponding TODO.md line.
+_ensure_todo_entry_written() {
+	local task_id="$1"
+	local issue_num="$2"
+	local description="$3"
+	local labels="$4"
+	local repo_path="$5"
+
+	local todo_file="${repo_path}/TODO.md"
+	[[ -f "$todo_file" ]] || return 0
+	[[ -n "$task_id" && -n "$issue_num" ]] || return 0
+
+	local task_id_ere
+	if declare -F _escape_ere >/dev/null 2>&1; then
+		task_id_ere=$(_escape_ere "$task_id")
+	else
+		task_id_ere="$task_id"
+	fi
+
+	# Fast path: entry already exists. Delegate to add_gh_ref_to_todo
+	# which idempotently stamps the ref:GH#NNN if missing.
+	if grep -qE "^[[:space:]]*- \[.\] ${task_id_ere}( |$)" "$todo_file"; then
+		if declare -F add_gh_ref_to_todo >/dev/null 2>&1; then
+			add_gh_ref_to_todo "$task_id" "$issue_num" "$todo_file" 2>/dev/null || true
+		fi
+		return 0
+	fi
+
+	# Build tag suffix from labels (skip reserved-prefix labels applied
+	# server-side by issue-sync / pulse, not authored in TODO).
+	local tags_str=""
+	if [[ -n "$labels" ]]; then
+		local _saved_ifs="$IFS"
+		IFS=','
+		local label
+		for label in $labels; do
+			label="${label#"${label%%[![:space:]]*}"}"
+			label="${label%"${label##*[![:space:]]}"}"
+			[[ -z "$label" ]] && continue
+			case "$label" in
+			status:* | tier:* | origin:* | dispatched:* | implemented:* | aidevops:*)
+				continue ;;
+			bug) tags_str="${tags_str:+${tags_str} }#bug" ;;
+			enhancement) tags_str="${tags_str:+${tags_str} }#feat" ;;
+			*) tags_str="${tags_str:+${tags_str} }#${label}" ;;
+			esac
+		done
+		IFS="$_saved_ifs"
+	fi
+
+	# Build the TODO line.
+	local safe_desc
+	safe_desc=$(printf '%s' "$description" | tr '\n\t' '  ' | sed -E 's/[[:space:]]+/ /g; s/^ //; s/ $//')
+	[[ -z "$safe_desc" ]] && safe_desc="(no description)"
+	local todo_line="- [ ] ${task_id} ${safe_desc}"
+	[[ -n "$tags_str" ]] && todo_line="${todo_line} ${tags_str}"
+	todo_line="${todo_line} ref:GH#${issue_num}"
+
+	_insert_todo_line "$todo_file" "$todo_line"
+
+	if declare -F log_info >/dev/null 2>&1; then
+		log_info "Appended TODO entry for ${task_id} (ref:GH#${issue_num})"
+	fi
+	return 0
+}
+
 # Create GitHub issue (post-allocation, non-blocking)
 # t1324: Delegates to issue-sync-helper.sh push when available for rich
 # issue bodies, proper labels (including auto-dispatch), and duplicate
 # detection. Falls back to bare gh issue create if helper not found.
+# t2548: Guarantees TODO.md entry after verified issue creation on both paths.
 create_github_issue() {
 	local title="$1"
 	local description="$2"
 	local labels="$3"
 	local repo_path="$4"
 
+	# t2548: extract task_id once for TODO-entry writes on both paths.
+	local _task_id_for_todo=""
+	[[ "$title" =~ ^(t[0-9]+(\.[0-9]+)*) ]] && _task_id_for_todo="${BASH_REMATCH[1]}"
+
 	cd "$repo_path" || return 1
 
 	# t2442: resolve repo slug once — used both by the delegation path
@@ -1228,6 +1376,11 @@ create_github_issue() {
 		_auto_assign_issue "$issue_num" "$repo_path"
 		_interactive_session_auto_claim_new_task "$issue_num" "$repo_path"
 		_lock_maintainer_issue_at_creation "$issue_num" "$repo_path"
+		# t2548: ensure TODO.md has the entry. On the delegation path
+		# push's `_push_process_task` returns SKIPPED silently when the
+		# task line is missing — issue is created but TODO never written.
+		# This call closes that gap idempotently.
+		_ensure_todo_entry_written "$_task_id_for_todo" "$issue_num" "$description" "$labels" "$repo_path"
 		# t2442: warn if parent-task label applied but body has no markers.
 		# The delegation path creates the issue via issue-sync-helper.sh
 		# cmd_push which ALREADY fires this warn — so we skip here to
@@ -1315,6 +1468,12 @@ create_github_issue() {
 		fi
 	fi
 
+	# t2548: ensure TODO.md has an entry for this task. The bare-fallback
+	# path creates the issue via `gh issue create` directly and never goes
+	# through issue-sync-helper.sh, so _push_process_task never runs and
+	# the TODO entry is never written. This call closes that gap idempotently.
+	_ensure_todo_entry_written "$_task_id_for_todo" "$issue_num" "$description" "$labels" "$repo_path"
+
 	echo "$issue_num"
 	return 0
 }
diff --git a/.agents/scripts/tests/test-claim-task-id-no-orphan.sh b/.agents/scripts/tests/test-claim-task-id-no-orphan.sh
new file mode 100755
index 0000000000..24cc94c6a6
--- /dev/null
+++ b/.agents/scripts/tests/test-claim-task-id-no-orphan.sh
@@ -0,0 +1,299 @@
+#!/usr/bin/env bash
+# SPDX-License-Identifier: MIT
+# SPDX-FileCopyrightText: 2025-2026 Marcus Quinn
+#
+# test-claim-task-id-no-orphan.sh — t2548 regression guard.
+#
+# Asserts that `_ensure_todo_entry_written()` in claim-task-id.sh appends a
+# TODO.md entry with ref:GH#NNN after a verified issue creation when the
+# task ID is not yet present in TODO.md.
+#
+# Production failure (t2548):
+#   `claim-task-id.sh --title X --description Y` created a GitHub issue
+#   (via _try_issue_sync_delegation OR the bare `gh issue create` fallback)
+#   but NEVER wrote a TODO.md entry when the task ID wasn't already
+#   registered in TODO.md. `_push_process_task` in issue-sync-helper.sh
+#   silently returns 0 when `target_task` is missing from TODO.md;
+#   `add_gh_ref_to_todo` is a modify-existing-line helper and no-ops.
+#   Result: issue created on GitHub, counter bumped, TODO entry never
+#   written. Evidence: 4/4 orphans from one ILDS session 2026-04-20
+#   (GH#555-558, t135-t138). johnwaldo/ilds had 9 open orphans in the
+#   corpus.
+#
+# Fix (Option A, t2548): idempotent `_ensure_todo_entry_written` helper
+# called from `create_github_issue` on both the delegation success path
+# and the bare-fallback path. If TODO.md already has the entry, delegates
+# to `add_gh_ref_to_todo` to stamp the ref; otherwise appends
+# `- [ ] <task_id> <description> <tags> ref:GH#<num>` to the `## Backlog`
+# section (EOF fallback if that heading is absent).
+#
+# Tests:
+#   1. Missing entry + no Backlog → appended at EOF with ref
+#   2. Missing entry + Backlog section present → appended inside Backlog
+#   3. Entry already present → ref stamped, no duplicate added
+#   4. Empty labels → no tag suffix, still valid line
+#   5. Reserved-prefix labels (status:, tier:, origin:) are skipped
+#   6. `bug` / `enhancement` labels map to `#bug` / `#feat` tags
+#   7. No TODO.md file → silent no-op (non-fatal)
+#
+# Cross-references: t2548 (this fix), GH#18352 (dispatch-dedup contract).
+
+set -uo pipefail
+
+SCRIPT_DIR_TEST="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)" || exit 1
+SCRIPTS_DIR="$(cd "${SCRIPT_DIR_TEST}/.." && pwd)" || exit 1
+
+if [[ -t 1 ]]; then
+	TEST_GREEN=$'\033[0;32m'
+	TEST_RED=$'\033[0;31m'
+	TEST_BLUE=$'\033[0;34m'
+	TEST_NC=$'\033[0m'
+else
+	TEST_GREEN="" TEST_RED="" TEST_BLUE="" TEST_NC=""
+fi
+
+TESTS_RUN=0
+TESTS_FAILED=0
+
+pass() {
+	local msg="$1"
+	TESTS_RUN=$((TESTS_RUN + 1))
+	printf '  %sPASS%s %s\n' "$TEST_GREEN" "$TEST_NC" "$msg"
+	return 0
+}
+
+fail() {
+	local msg="$1"
+	local detail="${2:-}"
+	TESTS_RUN=$((TESTS_RUN + 1))
+	TESTS_FAILED=$((TESTS_FAILED + 1))
+	printf '  %sFAIL%s %s\n' "$TEST_RED" "$TEST_NC" "$msg"
+	if [[ -n "$detail" ]]; then
+		printf '       %s\n' "$detail"
+	fi
+	return 0
+}
+
+# =============================================================================
+# Sandbox setup
+# =============================================================================
+TMP=$(mktemp -d -t t2548.XXXXXX)
+trap 'rm -rf "$TMP"' EXIT
+
+# Stub the log/print helpers used by the script so sourcing is quiet.
+# shellcheck disable=SC2317
+print_info() { return 0; }
+# shellcheck disable=SC2317
+print_warning() { return 0; }
+# shellcheck disable=SC2317
+print_error() { return 0; }
+# shellcheck disable=SC2317
+print_success() { return 0; }
+# shellcheck disable=SC2317
+log_verbose() { return 0; }
+# shellcheck disable=SC2317
+log_info() { return 0; }
+# shellcheck disable=SC2317
+log_warn() { return 0; }
+# shellcheck disable=SC2317
+log_error() { return 0; }
+# shellcheck disable=SC2317
+log_success() { return 0; }
+export -f print_info print_warning print_error print_success \
+	log_verbose log_info log_warn log_error log_success
+
+# shellcheck source=../claim-task-id.sh
+source "${SCRIPTS_DIR}/claim-task-id.sh" >/dev/null 2>&1 || true
+
+# Sanity: the helper must exist after sourcing.
+if ! declare -F _ensure_todo_entry_written >/dev/null 2>&1; then
+	printf '%sFATAL%s _ensure_todo_entry_written not defined after sourcing claim-task-id.sh\n' \
+		"$TEST_RED" "$TEST_NC"
+	exit 1
+fi
+
+printf '%sRunning _ensure_todo_entry_written tests (t2548)%s\n' \
+	"$TEST_BLUE" "$TEST_NC"
+
+# =============================================================================
+# Test 1 — Missing entry + no Backlog section → appended at EOF
+# =============================================================================
+REPO1="${TMP}/repo1"
+mkdir -p "$REPO1"
+cat >"${REPO1}/TODO.md" <<'EOF'
+# Project TODO
+
+## Ready
+
+- [ ] t001 pre-existing ref:GH#1
+
+## Done
+
+- [x] t000 historical ref:GH#0
+EOF
+
+_ensure_todo_entry_written "t2548" "20180" "fix orphan bug" "bug,auto-dispatch,tier:standard" "$REPO1"
+
+if grep -qE "^- \[ \] t2548 fix orphan bug .*ref:GH#20180" "${REPO1}/TODO.md"; then
+	pass "missing entry + no Backlog → appended at EOF with ref:GH#20180"
+else
+	fail "missing entry + no Backlog → appended at EOF with ref:GH#20180" \
+		"TODO.md after:\n$(cat "${REPO1}/TODO.md")"
+fi
+
+# Tier/status/origin labels skipped; auto-dispatch & bug become tags.
+if grep -qE "^- \[ \] t2548 fix orphan bug #bug #auto-dispatch ref:GH#20180$" "${REPO1}/TODO.md"; then
+	pass "reserved-prefix labels (tier:/status:/origin:) filtered from tag list"
+else
+	fail "reserved-prefix labels filtered" \
+		"line: $(grep 't2548' "${REPO1}/TODO.md" || echo '(no match)')"
+fi
+
+# =============================================================================
+# Test 2 — Missing entry + Backlog section → appended inside Backlog
+# ===================================
... (truncated, full diff at PR #20184)
```

</details>

--- a/.agents/scripts/backfill-orphan-todos.sh
+++ b/.agents/scripts/backfill-orphan-todos.sh
@@ -1,0 +1,220 @@
+#!/usr/bin/env bash
+# SPDX-License-Identifier: MIT
+# SPDX-FileCopyrightText: 2025-2026 Marcus Quinn
+#
+# backfill-orphan-todos.sh — one-shot repair tool for the t2548 orphan bug.
+#
+# Scans a repo's GitHub issues for "orphans" — issues whose title starts
+# with `tNNN: ` but whose `tNNN` ID has NO matching entry in TODO.md.
+# These are pre-fix leakage from `claim-task-id.sh` creating issues
+# without writing TODO.md (fixed in t2548 / GH#20180).
+#
+# Behaviour:
+#   list     (default) — print each orphan as `tNNN<TAB>GH#<num><TAB>title`
+#   annotate — append a `- [ ] tNNN <title> ref:GH#<num>` entry to the
+#              `## Backlog` section of TODO.md for every open orphan.
+#              Idempotent: skips IDs already in TODO.md.
+#              WRITES TODO.md in place; run inside a worktree, commit, PR.
+#
+# Scope:
+#   - Open issues only by default. Use `--include-closed` to also
+#     annotate orphans from closed issues (audit-trail completeness).
+#   - Only issues whose title matches `^t[0-9]+: ` are considered orphans.
+#
+# Usage:
+#   backfill-orphan-todos.sh list [--repo-path <path>] [--include-closed]
+#   backfill-orphan-todos.sh annotate [--repo-path <path>] [--include-closed] [--dry-run]
+#
+# Exit codes:
+#   0 — success (or no orphans found)
+#   1 — error (missing gh, no TODO.md, etc.)
+#
+# Cross-references:
+#   - t2548: root-cause fix (claim-task-id.sh::_ensure_todo_entry_written)
+
+set -uo pipefail
+
+SCRIPT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)" || exit 1
+
+# shellcheck source=shared-constants.sh
+if [[ -f "${SCRIPT_DIR}/shared-constants.sh" ]]; then
+	source "${SCRIPT_DIR}/shared-constants.sh"
+fi
+
+# Fallback helpers when shared-constants.sh is unavailable.
+if ! declare -F print_info >/dev/null 2>&1; then
+	print_info() { printf '[INFO] %s\n' "$*" >&2; return 0; }
+fi
+if ! declare -F print_warning >/dev/null 2>&1; then
+	print_warning() { printf '[WARN] %s\n' "$*" >&2; return 0; }
+fi
+if ! declare -F print_error >/dev/null 2>&1; then
+	print_error() { printf '[ERROR] %s\n' "$*" >&2; return 0; }
+fi
+
+usage() {
+	sed -n '7,30p' "$0" | sed 's/^# \{0,1\}//'
+	return 0
+}
+
+COMMAND="list"
+REPO_PATH=""
+INCLUDE_CLOSED=false
+DRY_RUN=false
+
+_parse_args() {
+	local cmd="${1:-list}"
+	COMMAND="$cmd"
+	shift || true
+	while [[ $# -gt 0 ]]; do
+		local flag="$1"; shift
+		case "$flag" in
+		--repo-path)
+			local _rp_val="$1"; shift
+			REPO_PATH="$_rp_val" ;;
+		--include-closed)
+			INCLUDE_CLOSED=true ;;
+		--dry-run)
+			DRY_RUN=true ;;
+		-h|--help)
+			usage; exit 0 ;;
+		*)
+			print_error "unknown flag: $flag"; usage; exit 1 ;;
+		esac
+	done
+	return 0
+}
+_parse_args "$@"
+
+[[ -z "$REPO_PATH" ]] && REPO_PATH="$(pwd)"
+
+if ! command -v gh >/dev/null 2>&1; then
+	print_error "gh CLI not found"; exit 1
+fi
+
+TODO_FILE="${REPO_PATH}/TODO.md"
+if [[ ! -f "$TODO_FILE" ]]; then
+	print_error "no TODO.md at ${REPO_PATH}"; exit 1
+fi
+
+SLUG=$(git -C "$REPO_PATH" remote get-url origin 2>/dev/null \
+	| sed 's|.*github\.com[:/]||;s|\.git$||' || true)
+if [[ -z "$SLUG" ]]; then
+	print_error "could not resolve GitHub slug from origin remote in ${REPO_PATH}"
+	exit 1
+fi
+
+STATE_FILTER="open"
+[[ "$INCLUDE_CLOSED" == "true" ]] && STATE_FILTER="all"
+
+print_info "scanning ${SLUG} (${STATE_FILTER}) for tNNN: orphans..."
+
+ISSUES_JSON=$(gh issue list --repo "$SLUG" --state "$STATE_FILTER" \
+	--search "t in:title" --limit 1000 \
+	--json number,title 2>/dev/null || echo "[]")
+
+CANDIDATES=$(printf '%s' "$ISSUES_JSON" | jq -r '
+	.[]
+	| select(.title | test("^t[0-9]+:"))
+	| [
+		(.title | capture("^(?<id>t[0-9]+):") | .id),
+		(.number | tostring),
+		.title
+	]
+	| @tsv
+' 2>/dev/null || echo "")
+
+if [[ -z "$CANDIDATES" ]]; then
+	print_info "no tNNN-prefixed issues found"; exit 0
+fi
+
+declare -a ORPHANS=()
+
+_find_orphans() {
+	local task_id issue_num title
+	while IFS=$'\t' read -r task_id issue_num title; do
+		[[ -z "$task_id" || -z "$issue_num" ]] && continue
+		if grep -qE "^[[:space:]]*- \[.\] ${task_id}( |$)" "$TODO_FILE"; then
+			continue
+		fi
+		ORPHANS+=("${task_id}"$'\t'"${issue_num}"$'\t'"${title}")
+	done <<<"$CANDIDATES"
+	return 0
+}
+_find_orphans
+
+ORPHAN_COUNT=${#ORPHANS[@]}
+
+if [[ "$ORPHAN_COUNT" -eq 0 ]]; then
+	print_info "no orphans found in ${SLUG}"; exit 0
+fi
+
+_annotate_orphans() {
+	print_info "${ORPHAN_COUNT} orphan(s) to annotate in ${TODO_FILE}"
+	if [[ "$DRY_RUN" == "true" ]]; then
+		print_info "[DRY-RUN] would append:"
+		local row task_id issue_num title desc
+		for row in "${ORPHANS[@]}"; do
+			IFS=$'\t' read -r task_id issue_num title <<<"$row"
+			desc="${title#"${task_id}": }"
+			printf '  - [ ] %s %s ref:GH#%s\n' "$task_id" "$desc" "$issue_num"
+		done
+		return 0
+	fi
+
+	local backlog_start
+	backlog_start=$(grep -nE '^## Backlog[[:space:]]*$' "$TODO_FILE" 2>/dev/null \
+		| head -1 | cut -d: -f1 || true)
+
+	local tmp
+	tmp=$(mktemp)
+	trap 'rm -f "$tmp"' RETURN
+
+	local row task_id issue_num title desc
+	for row in "${ORPHANS[@]}"; do
+		IFS=$'\t' read -r task_id issue_num title <<<"$row"
+		desc="${title#"${task_id}": }"
+		local todo_line="- [ ] ${task_id} ${desc} ref:GH#${issue_num}"
+
+		if [[ -n "$backlog_start" ]]; then
+			local next_heading
+			next_heading=$(awk -v s="$backlog_start" \
+				'NR > s && /^## / {print NR; exit}' "$TODO_FILE" 2>/dev/null || true)
+			if [[ -z "$next_heading" ]]; then
+				cat "$TODO_FILE" >"$tmp"
+				[[ -n "$(tail -c 1 "$tmp")" ]] && printf '\n' >>"$tmp"
+				printf '%s\n' "$todo_line" >>"$tmp"
+			else
+				awk -v nh="$next_heading" -v line="$todo_line" '
+					NR == nh { print line; print "" }
+					{ print }
+				' "$TODO_FILE" >"$tmp"
+			fi
+		else
+			cat "$TODO_FILE" >"$tmp"
+			printf '\n## Backlog (t2548 backfill)\n\n%s\n' "$todo_line" >>"$tmp"
+		fi
+
+		[[ -s "$tmp" ]] && mv "$tmp" "$TODO_FILE"
+		# Recompute backlog_start for the next iteration (file grew).
+		backlog_start=$(grep -nE '^## Backlog' "$TODO_FILE" 2>/dev/null \
+			| head -1 | cut -d: -f1 || true)
+	done
+
+	print_info "annotated ${ORPHAN_COUNT} orphan(s) — review 'git diff' and commit"
+	return 0
+}
+
+case "$COMMAND" in
+list)
+	print_info "${ORPHAN_COUNT} orphan(s) in ${SLUG}:"
+	printf '%s\n' "${ORPHANS[@]}"
+	exit 0
+	;;
+annotate)
+	_annotate_orphans
+	exit 0
+	;;
+*)
+	print_error "unknown command: ${COMMAND}"; usage; exit 1 ;;
+esac

--- a/.agents/scripts/claim-task-id.sh
+++ b/.agents/scripts/claim-task-id.sh
@@ -1207,10 +1207,131 @@ _compose_issue_body() {
 	return 0
 }
 
+# _insert_todo_line FILE LINE
+# Write LINE into FILE at the end of the ## Backlog section (or at EOF).
+# Extracted from _ensure_todo_entry_written to keep function bodies <=100 lines.
+_insert_todo_line() {
+	local todo_file="$1"
+	local todo_line="$2"
+
+	local backlog_start
+	backlog_start=$(grep -nE '^## Backlog[[:space:]]*$' "$todo_file" 2>/dev/null \
+		| head -1 | cut -d: -f1 || true)
+
+	local tmp
+	tmp=$(mktemp)
+	if [[ -n "$backlog_start" ]]; then
+		local next_heading
+		next_heading=$(awk -v s="$backlog_start" \
+			'NR > s && /^## / {print NR; exit}' "$todo_file" 2>/dev/null || true)
+		if [[ -z "$next_heading" ]]; then
+			# Backlog is the last section — insert before trailing blank lines.
+			awk -v line="$todo_line" '
+				{ lines[NR] = $0 }
+				END {
+					last = NR
+					while (last > 0 && lines[last] == "") last--
+					for (i = 1; i <= last; i++) print lines[i]
+					print line
+					for (i = last + 1; i <= NR; i++) print lines[i]
+				}
+			' "$todo_file" >"$tmp"
+		else
+			# Insert just before the next heading.
+			awk -v nh="$next_heading" -v line="$todo_line" '
+				NR == nh { print line; print "" }
+				{ print }
+			' "$todo_file" >"$tmp"
+		fi
+	else
+		# No Backlog section — append at EOF.
+		cat "$todo_file" >"$tmp"
+		printf '\n%s\n' "$todo_line" >>"$tmp"
+	fi
+
+	if [[ -s "$tmp" ]]; then
+		mv "$tmp" "$todo_file"
+	else
+		rm -f "$tmp"
+	fi
+	return 0
+}
+
+# _ensure_todo_entry_written TASK_ID ISSUE_NUM DESCRIPTION LABELS REPO_PATH
+# t2548: Idempotently appends a TODO.md entry after verified GitHub issue
+# creation. Closes the orphan gap where both create_github_issue() paths
+# created issues without writing a corresponding TODO.md line.
+#
+# - If TODO.md already has a matching entry, delegates to add_gh_ref_to_todo
+#   to stamp the ref:GH#NNN if missing (idempotent).
+# - Otherwise appends `- [ ] <task_id> <description> <tags> ref:GH#<num>`
+#   to the `## Backlog` section (falls back to EOF if absent).
+# - Labels with status:, tier:, origin:, dispatched:, implemented: prefixes
+#   are skipped — they are not TODO-file-format tags.
+# Returns 0 always (non-fatal; the issue is already created).
+_ensure_todo_entry_written() {
+	local task_id="$1"
+	local issue_num="$2"
+	local description="$3"
+	local labels="$4"
+	local repo_path="$5"
+
+	local todo_file="${repo_path}/TODO.md"
+	[[ -f "$todo_file" ]] || return 0
+	[[ -n "$task_id" && -n "$issue_num" ]] || return 0
+
+	# Fast path: entry already exists — stamp the ref if missing.
+	if grep -qE "^[[:space:]]*- \[.\] ${task_id}( |$)" "$todo_file"; then
+		if declare -F add_gh_ref_to_todo >/dev/null 2>&1; then
+			add_gh_ref_to_todo "$task_id" "$issue_num" "$todo_file" 2>/dev/null || true
+		fi
+		return 0
+	fi
+
+	# Build tag suffix from labels (skip reserved-prefix labels applied
+	# server-side by issue-sync / pulse, not authored in TODO).
+	local tags_str=""
+	if [[ -n "$labels" ]]; then
+		local _saved_ifs="$IFS"
+		IFS=','
+		local label
+		for label in $labels; do
+			label="${label#"${label%%[![:space:]]*}"}"
+			label="${label%"${label##*[![:space:]]}"}"
+			[[ -z "$label" ]] && continue
+			case "$label" in
+			status:* | tier:* | origin:* | dispatched:* | implemented:* | aidevops:*)
+				continue ;;
+			bug)         tags_str="${tags_str:+${tags_str} }#bug" ;;
+			enhancement) tags_str="${tags_str:+${tags_str} }#feat" ;;
+			*)           tags_str="${tags_str:+${tags_str} }#${label}" ;;
+			esac
+		done
+		IFS="$_saved_ifs"
+	fi
+
+	# Build the TODO line.
+	local safe_desc
+	safe_desc=$(printf '%s' "$description" \
+		| tr '\n\t' '  ' | sed -E 's/[[:space:]]+/ /g; s/^ //; s/ $//')
+	[[ -z "$safe_desc" ]] && safe_desc="(no description)"
+	local todo_line="- [ ] ${task_id} ${safe_desc}"
+	[[ -n "$tags_str" ]] && todo_line="${todo_line} ${tags_str}"
+	todo_line="${todo_line} ref:GH#${issue_num}"
+
+	_insert_todo_line "$todo_file" "$todo_line"
+
+	if declare -F log_info >/dev/null 2>&1; then
+		log_info "t2548: appended TODO entry for ${task_id} (ref:GH#${issue_num})"
+	fi
+	return 0
+}
+
 # Create GitHub issue (post-allocation, non-blocking)
 # t1324: Delegates to issue-sync-helper.sh push when available for rich
 # issue bodies, proper labels (including auto-dispatch), and duplicate
 # detection. Falls back to bare gh issue create if helper not found.
+# t2548: Guarantees TODO.md entry after verified issue creation on both paths.
 create_github_issue() {
 	local title="$1"
 	local description="$2"
@@ -1218,6 +1339,11 @@ create_github_issue() {
 	local repo_path="$4"
 
 	cd "$repo_path" || return 1
+
+	# t2548: extract task_id once — used on both creation paths to write
+	# the TODO.md entry after verified issue creation.
+	local _task_id_for_todo=""
+	[[ "$title" =~ ^(t[0-9]+(\.[0-9]+)*) ]] && _task_id_for_todo="${BASH_REMATCH[1]}"
 
 	# t2442: resolve repo slug once — used both by the delegation path
 	# and the bare-fallback path for the parent-task warn call. Must run
@@ -1232,6 +1358,12 @@ create_github_issue() {
 		_auto_assign_issue "$issue_num" "$repo_path"
 		_interactive_session_auto_claim_new_task "$issue_num" "$repo_path"
 		_lock_maintainer_issue_at_creation "$issue_num" "$repo_path"
+		# t2548: ensure TODO.md has the entry. On the delegation path
+		# issue-sync-helper.sh _push_process_task silently returns SKIPPED
+		# when the task line is absent from TODO.md — issue is created but
+		# TODO never written. This call closes that gap idempotently.
+		_ensure_todo_entry_written \
+			"$_task_id_for_todo" "$issue_num" "$description" "$labels" "$repo_path"
 		# t2442: warn if parent-task label applied but body has no markers.
 		# The delegation path creates the issue via issue-sync-helper.sh
 		# cmd_push which ALREADY fires this warn — so we skip here to
@@ -1318,6 +1450,13 @@ create_github_issue() {
 			fi
 		fi
 	fi
+
+	# t2548: ensure TODO.md has an entry for this task. The bare-fallback
+	# path creates the issue via `gh issue create` directly and never goes
+	# through issue-sync-helper.sh, so _push_process_task never runs and
+	# the TODO entry is never written. This call closes that gap idempotently.
+	_ensure_todo_entry_written \
+		"$_task_id_for_todo" "$issue_num" "$description" "$labels" "$repo_path"
 
 	echo "$issue_num"
 	return 0

--- a/.agents/scripts/tests/test-claim-task-id-no-orphan.sh
+++ b/.agents/scripts/tests/test-claim-task-id-no-orphan.sh
@@ -1,0 +1,316 @@
+#!/usr/bin/env bash
+# SPDX-License-Identifier: MIT
+# SPDX-FileCopyrightText: 2025-2026 Marcus Quinn
+#
+# test-claim-task-id-no-orphan.sh — t2548 regression guard.
+#
+# Asserts that `_ensure_todo_entry_written()` in claim-task-id.sh appends a
+# TODO.md entry with ref:GH#NNN after a verified issue creation when the
+# task ID is not yet present in TODO.md.
+#
+# Production failure (t2548):
+#   `claim-task-id.sh --title X --description Y` created a GitHub issue
+#   (via _try_issue_sync_delegation OR the bare `gh issue create` fallback)
+#   but NEVER wrote a TODO.md entry when the task ID wasn't already
+#   registered in TODO.md. `_push_process_task` in issue-sync-helper.sh
+#   silently returns 0 when `target_task` is missing from TODO.md;
+#   `add_gh_ref_to_todo` is a modify-existing-line helper and no-ops.
+#   Evidence: 4/4 orphans from one ILDS session 2026-04-20 (t135-t138).
+#
+# Fix (Option A, t2548): idempotent `_ensure_todo_entry_written` helper
+# called from `create_github_issue` on both the delegation success path
+# and the bare-fallback path.
+#
+# Tests:
+#   1. Missing entry + no Backlog → appended at EOF with ref
+#   2. Missing entry + Backlog section present → appended inside Backlog
+#   3. Entry already present → ref stamped, no duplicate added
+#   4. Empty labels → no tag suffix, still valid line
+#   5. Reserved-prefix labels (status:, tier:, origin:) are skipped
+#   6. `bug` / `enhancement` labels map to `#bug` / `#feat` tags
+#   7. No TODO.md file → silent no-op (non-fatal)
+#   8. Empty task_id or issue_num → silent no-op (non-fatal)
+
+set -u
+
+SCRIPT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)" || exit 1
+CLAIM_SCRIPT="${SCRIPT_DIR}/../claim-task-id.sh"
+
+RED=$'\033[0;31m'
+GREEN=$'\033[0;32m'
+BLUE=$'\033[0;34m'
+NC=$'\033[0m'
+
+PASS=0
+FAIL=0
+ERRORS=""
+
+pass() {
+	local name="$1"
+	printf '%s[PASS]%s %s\n' "$GREEN" "$NC" "$name"
+	PASS=$((PASS + 1))
+	return 0
+}
+
+fail() {
+	local name="$1"
+	local detail="${2:-}"
+	printf '%s[FAIL]%s %s\n' "$RED" "$NC" "$name"
+	[[ -n "$detail" ]] && printf '       %s\n' "$detail"
+	FAIL=$((FAIL + 1))
+	ERRORS="${ERRORS}\n  - ${name}: ${detail}"
+	return 0
+}
+
+# Source claim-task-id.sh to gain access to internal helper functions.
+# The BASH_SOURCE guard in the script prevents main() from running.
+# shellcheck disable=SC1090
+if ! source "$CLAIM_SCRIPT" 2>/dev/null; then
+	printf '%s[FATAL]%s Failed to source %s\n' "$RED" "$NC" "$CLAIM_SCRIPT" >&2
+	exit 1
+fi
+
+if ! declare -F _ensure_todo_entry_written >/dev/null 2>&1; then
+	printf '%s[FATAL]%s _ensure_todo_entry_written not defined after sourcing\n' \
+		"$RED" "$NC" >&2
+	exit 1
+fi
+
+printf '%sRunning _ensure_todo_entry_written tests (t2548)%s\n' "$BLUE" "$NC"
+
+# ---------------------------------------------------------------------------
+# Test 1 — Missing entry + no Backlog section → appended at EOF
+# ---------------------------------------------------------------------------
+test_missing_no_backlog() {
+	local name="1: missing entry + no Backlog → appended at EOF with ref"
+	local tmpdir
+	tmpdir=$(mktemp -d)
+	# shellcheck disable=SC2064
+	trap "rm -rf '$tmpdir'" RETURN
+
+	cat >"${tmpdir}/TODO.md" <<'EOF'
+# Project TODO
+
+## Ready
+
+- [ ] t001 pre-existing ref:GH#1
+
+## Done
+
+- [x] t000 historical ref:GH#0
+EOF
+
+	_ensure_todo_entry_written "t2548" "20180" "fix orphan bug" "" "$tmpdir"
+
+	if grep -qE "^- \[ \] t2548 fix orphan bug ref:GH#20180$" "${tmpdir}/TODO.md"; then
+		pass "$name"
+	else
+		fail "$name" "line not found: $(grep 't2548' "${tmpdir}/TODO.md" || echo '(none)')"
+	fi
+	return 0
+}
+test_missing_no_backlog
+
+# ---------------------------------------------------------------------------
+# Test 2 — Missing entry + Backlog section → appended inside Backlog
+# ---------------------------------------------------------------------------
+test_missing_with_backlog() {
+	local name="2: missing entry + Backlog present → appended inside Backlog"
+	local tmpdir
+	tmpdir=$(mktemp -d)
+	# shellcheck disable=SC2064
+	trap "rm -rf '$tmpdir'" RETURN
+
+	cat >"${tmpdir}/TODO.md" <<'EOF'
+# Project TODO
+
+## Tasks
+
+- [ ] t001 existing task ref:GH#1
+
+## Backlog
+
+- [ ] t002 backlog item ref:GH#2
+
+## Done
+
+- [x] t000 done ref:GH#0
+EOF
+
+	_ensure_todo_entry_written "t2548" "20180" "fix orphan bug" "" "$tmpdir"
+
+	# Line must exist AND appear between ## Backlog and ## Done
+	local backlog_line done_line new_line
+	backlog_line=$(grep -n '^## Backlog' "${tmpdir}/TODO.md" | cut -d: -f1)
+	done_line=$(grep -n '^## Done' "${tmpdir}/TODO.md" | cut -d: -f1)
+	new_line=$(grep -n 't2548' "${tmpdir}/TODO.md" | cut -d: -f1)
+
+	if [[ -n "$new_line" && -n "$backlog_line" && -n "$done_line" ]] \
+		&& [[ "$new_line" -gt "$backlog_line" ]] \
+		&& [[ "$new_line" -lt "$done_line" ]]; then
+		pass "$name"
+	else
+		fail "$name" "backlog=${backlog_line} done=${done_line} new=${new_line:-absent}"
+	fi
+	return 0
+}
+test_missing_with_backlog
+
+# ---------------------------------------------------------------------------
+# Test 3 — Entry already present → no duplicate line added
+# ---------------------------------------------------------------------------
+test_entry_already_present() {
+	local name="3: entry already present → no duplicate line added"
+	local tmpdir
+	tmpdir=$(mktemp -d)
+	# shellcheck disable=SC2064
+	trap "rm -rf '$tmpdir'" RETURN
+
+	cat >"${tmpdir}/TODO.md" <<'EOF'
+# Tasks
+
+- [ ] t2548 fix orphan bug ref:GH#20180
+EOF
+
+	_ensure_todo_entry_written "t2548" "20180" "fix orphan bug" "" "$tmpdir"
+
+	local count
+	count=$(grep -c 't2548' "${tmpdir}/TODO.md" 2>/dev/null || echo 0)
+	if [[ "$count" -eq 1 ]]; then
+		pass "$name"
+	else
+		fail "$name" "expected 1 occurrence, got ${count}"
+	fi
+	return 0
+}
+test_entry_already_present
+
+# ---------------------------------------------------------------------------
+# Test 4 — Empty labels → no tag suffix, valid line
+# ---------------------------------------------------------------------------
+test_empty_labels() {
+	local name="4: empty labels → no tag suffix in appended line"
+	local tmpdir
+	tmpdir=$(mktemp -d)
+	# shellcheck disable=SC2064
+	trap "rm -rf '$tmpdir'" RETURN
+
+	printf '# Tasks\n\n' >"${tmpdir}/TODO.md"
+
+	_ensure_todo_entry_written "t2548" "20180" "fix orphan bug" "" "$tmpdir"
+
+	if grep -qE "^- \[ \] t2548 fix orphan bug ref:GH#20180$" "${tmpdir}/TODO.md"; then
+		pass "$name"
+	else
+		fail "$name" "$(grep 't2548' "${tmpdir}/TODO.md" || echo '(none)')"
+	fi
+	return 0
+}
+test_empty_labels
+
+# ---------------------------------------------------------------------------
+# Test 5 — Reserved-prefix labels skipped; plain labels become tags
+# ---------------------------------------------------------------------------
+test_reserved_label_filtering() {
+	local name="5: reserved-prefix labels filtered; plain labels become tags"
+	local tmpdir
+	tmpdir=$(mktemp -d)
+	# shellcheck disable=SC2064
+	trap "rm -rf '$tmpdir'" RETURN
+
+	printf '# Tasks\n\n' >"${tmpdir}/TODO.md"
+
+	_ensure_todo_entry_written "t2548" "20180" "fix orphan bug" \
+		"status:queued,tier:standard,origin:worker,auto-dispatch" "$tmpdir"
+
+	if grep -qE "^- \[ \] t2548 fix orphan bug #auto-dispatch ref:GH#20180$" \
+		"${tmpdir}/TODO.md"; then
+		pass "$name"
+	else
+		fail "$name" "$(grep 't2548' "${tmpdir}/TODO.md" || echo '(none)')"
+	fi
+	return 0
+}
+test_reserved_label_filtering
+
+# ---------------------------------------------------------------------------
+# Test 6 — `bug` / `enhancement` labels map to `#bug` / `#feat`
+# ---------------------------------------------------------------------------
+test_label_mapping() {
+	local name="6: bug/enhancement labels map to #bug/#feat"
+	local tmpdir
+	tmpdir=$(mktemp -d)
+	# shellcheck disable=SC2064
+	trap "rm -rf '$tmpdir'" RETURN
+
+	printf '# Tasks\n\n' >"${tmpdir}/TODO.md"
+
+	_ensure_todo_entry_written "t2548" "20180" "fix orphan bug" \
+		"bug,enhancement" "$tmpdir"
+
+	if grep -qE "^- \[ \] t2548 fix orphan bug #bug #feat ref:GH#20180$" \
+		"${tmpdir}/TODO.md"; then
+		pass "$name"
+	else
+		fail "$name" "$(grep 't2548' "${tmpdir}/TODO.md" || echo '(none)')"
+	fi
+	return 0
+}
+test_label_mapping
+
+# ---------------------------------------------------------------------------
+# Test 7 — No TODO.md file → silent no-op
+# ---------------------------------------------------------------------------
+test_no_todo_file() {
+	local name="7: no TODO.md → silent no-op (non-fatal)"
+	local tmpdir
+	tmpdir=$(mktemp -d)
+	# shellcheck disable=SC2064
+	trap "rm -rf '$tmpdir'" RETURN
+
+	local rc=0
+	_ensure_todo_entry_written "t2548" "20180" "fix orphan bug" "" "$tmpdir" || rc=$?
+	if [[ $rc -eq 0 ]]; then
+		pass "$name"
+	else
+		fail "$name" "returned non-zero: $rc"
+	fi
+	return 0
+}
+test_no_todo_file
+
+# ---------------------------------------------------------------------------
+# Test 8 — Empty task_id or issue_num → silent no-op
+# ---------------------------------------------------------------------------
+test_empty_task_id() {
+	local name="8: empty task_id → silent no-op"
+	local tmpdir
+	tmpdir=$(mktemp -d)
+	# shellcheck disable=SC2064
+	trap "rm -rf '$tmpdir'" RETURN
+
+	printf '# Tasks\n\n' >"${tmpdir}/TODO.md"
+
+	local rc=0
+	_ensure_todo_entry_written "" "20180" "fix orphan bug" "" "$tmpdir" || rc=$?
+	local lc
+	lc=$(wc -l <"${tmpdir}/TODO.md")
+	if [[ $rc -eq 0 && "$lc" -le 3 ]]; then
+		pass "$name"
+	else
+		fail "$name" "rc=${rc} lines=${lc}"
+	fi
+	return 0
+}
+test_empty_task_id
+
+# ---------------------------------------------------------------------------
+# Summary
+# ---------------------------------------------------------------------------
+printf '\n'
+printf '%sResults: %d passed, %d failed%s\n' "$BLUE" "$PASS" "$FAIL" "$NC"
+if [[ $FAIL -gt 0 ]]; then
+	printf '%sFailed tests:%s%s\n' "$RED" "$NC" "$ERRORS"
+	exit 1
+fi
+exit 0

